### PR TITLE
[openwrt] update the Makefile and init scripts

### PR DESCRIPTION
**Pull request by caipiblack**
Some modifications to make ot-br-posix working with OpenWRT.

Compatible with Matter (tested on TE9 and SDK 1.0.0)

This pull request is related to:

- This issue: https://github.com/openthread/ot-br-posix/issues/1250 
- This pull-request: https://github.com/openthread/ot-br-posix/pull/1410

I re-created the pull request to make it clearer.

### CODE REVIEW

**Comment by caipiblack** [comment]
This is how I have tested my changes, and how you can check.

Using ot-br-posix on OpenWRT
----------------------------

This document describe how to setup `ot-br-posix` on OpenWRT. For the example we are using a `WRT3200ACM` router.

## Install OpenWRT on WRT3200ACM

Go to the OpenWRT wiki page about this router and execute the instructions. [click here](https://openwrt.org/toh/linksys/wrt3200acm)

This example use the release `21.02.3` but you can use a more recent.

## SDK Setup

1. Go to the download page of OpenWRT for the target: `https://downloads.openwrt.org/releases/21.02.3/targets/mvebu/cortexa9/`.
2. Download the SDK `openwrt-sdk-21.02.3-mvebu-cortexa9_gcc-8.4.0_musl_eabi.Linux-x86_64.tar.xz`.
3. Extract the archive into a folder `sdk`.
4. Copy `feeds.conf.default` to `feeds.conf`.
5. Add a new line in `feeds.conf`: `src-link openthread /..path_to../ot-br-posix/etc/openwrt`.
6. Update the feeds and install `openthread-br`:
```
./scripts/feeds update -a
./scripts/feeds install openthread-br
```
7. Configure the SDK (optional):
```
make menuconfig
# Disable: Global build settings > Cryptographically sign package lists
# Disable: Advanced configuration options (for developers) > Automatic rebuild of packages
# Disable: Advanced configuration options (for developers) > Automatic removal of build directories
```
8. Build:
```
make defconfig
make -j
```
9. On the OpenWRT router, create a temporary folder to store the generated packages: `mkdir /tmp/pkg`.
10. Transfer the required packages to the router:
```
scp \
bin/packages/arm_cortex-a9_vfpv3-d16/openthread/openthread-br_1.0-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libdbus_1.13.12-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/dbus_1.13.12-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libdaemon_0.14-5_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libexpat_2.2.10-2_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libavahi-dbus-support_0.8-7_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/avahi-dbus-daemon_0.8-7_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libavahi-client_0.8-7_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/base/ipset_7.6-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/base/libmnl0_1.0.4-2_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/base/libipset13_7.6-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/librt_1.1.24-3_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/libstdcpp6_8.4.0-3_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/iptables-mod-extra_1.8.7-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-usb-acm_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-tun_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-ipt-extra_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-ipt-ipset_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-nfnetlink_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
root@192.168.3.1:/tmp/pkg
```
11. Install the packages:

```
opkg install /tmp/pkg/*
```

The `otbr-agent` and `otbr-firewall` serviced are automatically started after install and when the system boot.

## Test with Thread client

1. On the router, create the thread network:
```
ot-ctl factoryreset
ot-ctl dataset init new
ot-ctl dataset commit active
ot-ctl ifconfig up
ot-ctl thread start
ot-ctl dataset active -x
```
2. On a thread device with `ot-cli` firmware, Configure it to join the thread network:
```
dataset set active 0e080000000000010000000300001435060004001fffe00208531094376d6c809d0708fdf8a372297c3f2e0510a23392b82f3139418226700a4a26e242030f4f70656e5468726561642d31336630010213f00410f1d1802d25b3abb62219a3771b469cd90c0402a0f7f8
ifconfig up
thread start
ipaddr
```
3. Make sure you can ping the IPV6 of the thread device from the router.
4. Make sure you can ping the IPV6 of the router from the thread device.
5. Make sure you can ping the IPV6 of the thread device from your computer (the computer must be connected to a LAN port of the router).

_In case of failure at the point #5, use wireshark to check if your computer receive a ICMPv6 message with router advertisement when the thread network is created and make sure your computer accept it._

## Test with Matter device

For the test I am using a real device on `v1.0.0` tag.

1. Commission the device using `chip-tool`:

```
./chip-tool pairing code-thread 4 hex:0e080000000000010000000300001435060004001fffe00208531094376d6c809d0708fdf8a372297c3f2e0510a23392b82f3139418226700a4a26e242030f4f70656e5468726561642d31336630010213f00410f1d1802d25b3abb62219a3771b469cd90c0402a0f7f8 MT:-AM14A-E15HRI02VC10
```

[Logs of the commissioning](https://github.com/openthread/ot-br-posix/files/10272248/matter_commissioning.txt)

2. Control the device using `chip-tool`:

```
./chip-tool onoff toggle 4 1
```

[Logs of the toggle command](https://github.com/openthread/ot-br-posix/files/10272250/matter_control.txt)

**Comment by caipiblack** [comment]
**At this point:**

- INFRA_IF_NAME is configurable with UCI
- THREAD_IF_NAME is configurable with UCI
- The script otbr-firewall is fixed (in case of THREAD_IF_NAME changes while the script is running, we can still stop it and remove ipset rules)

**Remaining work:**

- (optionnal) : Make INFRA_IF_NAME and THREAD_IF_NAME also configurable from the web page. (need ui design ideas/instructions or someone to do it)
- In the web page there are references to "wpan0" statically (src/openwrt/view/admin_thread/thread_overview.htm), it should be good to display the correct interface name, depending on the interface used by otbr-agent. This point is not necessary related to this PR.

File: etc/openwrt/openthread-br/Makefile [hunk 1]
```diff
@@ -43,16 +43,23 @@ CMAKE_SOURCE_DIR=$(LOCAL_SOURCE_DIR)
 CMAKE_OPTIONS+= \
 	-DBUILD_TESTING=OFF \
 	-DCMAKE_INSTALL_PREFIX=/usr \
-	-DOTBR_BORDER_AGENT=OFF \
-	-DOTBR_MDNS=OFF \
+	-DOTBR_OPENWRT=ON \
```
Comment (bukepo) [inline etc/openwrt/openthread-br/Makefile:46]: I think we tried to keep all these options sorted.

File: etc/openwrt/openthread-br/Makefile [hunk 2]
```diff
@@ -43,16 +43,23 @@ CMAKE_SOURCE_DIR=$(LOCAL_SOURCE_DIR)
...
+	-DOT_FIREWALL=ON
+
+TARGET_CFLAGS += -DOPENTHREAD_POSIX_CONFIG_DAEMON_SOCKET_BASENAME=\\\"/var/run/openthread-%s\\\"
 
 define Package/openthread-br
 	SECTION:=base
 	CATEGORY:=Network
 	TITLE:=OpenThread Border Router
-	DEPENDS:=+libstdcpp +libjson-c +libubus +libblobmsg-json
+	DEPENDS:=+libstdcpp +libjson-c +libubus +libblobmsg-json +libavahi-client +iptables-mod-extra +ipset
```
Comment (bukepo) [inline etc/openwrt/openthread-br/Makefile:62]: should we add avahi-daemon as the runtime dependency as well?

Comment (caipiblack) [reply etc/openwrt/openthread-br/Makefile:62]: Yes, we have to! I will do it

**Review by caipiblack — COMMENTED** [review]
In my opinion, the objective of the OpenWRT implementation is to propose a base that should be customised during the final integration.

I revert my changes about backbone feature because it was not really activated (as the cmake options was disabled) and it's maybe not really usefull for everybody.

**Review by caipiblack — COMMENTED** [review]
Previously the thread interface name was fixed to "wpan0" in the web ui.

This commit is just to make it "dynamic" so it match with the interface name given at the begin (or default value)

image

File: src/openwrt/CMakeLists.txt [hunk 3]
```diff
@@ -32,3 +32,16 @@ configure_file(otbr-agent.init.in otbr-agent.init)
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/otbr-agent.init
     DESTINATION ${CMAKE_INSTALL_FULL_SYSCONFDIR}/init.d
     RENAME otbr-agent)
+
```
Comment (caipiblack) [inline src/openwrt/CMakeLists.txt:35]: Currently `otbr-agent.uci-config.in` and `otbr-firewall.init.in` pass through the cmake configure_file() but we don't have any "cmake variables" on these files.

idealy a variable for the default thread interface will be nice, so the default value could be configured at compile time (wpan0).

But it doesn't seems to be implemented yet in ot-br-posix.

File: src/openwrt/CMakeLists.txt [hunk 4]
```diff
@@ -32,3 +32,16 @@ configure_file(otbr-agent.init.in otbr-agent.init)
...
+    DESTINATION ${CMAKE_INSTALL_FULL_SYSCONFDIR}/config
+    RENAME otbr-agent)
+
+
+if(OT_FIREWALL)
+    configure_file(otbr-firewall.init.in otbr-firewall.init)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/otbr-firewall.init
+        DESTINATION ${CMAKE_INSTALL_FULL_SYSCONFDIR}/init.d
+        RENAME otbr-firewall)
+endif(OT_FIREWALL)
```
Comment (caipiblack) [inline src/openwrt/CMakeLists.txt:47]: Currently `otbr-agent.uci-config.in` and `otbr-firewall.init.in` pass through the cmake configure_file() but we don't have any "cmake variables" on these files.

idealy a variable for the default thread interface will be nice, so the default value could be configured at compile time (wpan0).

But it doesn't seems to be implemented yet in ot-br-posix.

**Review by bukepo — APPROVED** [review]

**Pull request merged**